### PR TITLE
Added more analysis SMB for IO timeouts and Dirty Shutdowns

### PR DESCRIPTION
**Pull request by girdharb**
Added more analysis SMB for IO timeouts and Dirty Shutdowns.
In the process re-factored the function to be reusable for other event as well.

### CODE REVIEW

File: PrivateCloud.DiagnosticInfo/PrivateCloud.DiagnosticInfo.psm1 [hunk 1]
```diff
@@ -5086,16 +5084,72 @@ WARNING: the SMB Client is receiving RDMA disconnects. This is an error whose ro
...
 WARNING: the SMB Client is receiving RDMA connect errors. This is an error whose root
 `t cause may be actual lack of connectivity or fundamental problems with the RDMA
 `t network fabric. Please inspect especially if in the Last5 bucket.
 "@
 
-    $j += Start-Job -name 'SMB Connectivity Error Check - Connect Failures (Event 30803)' -InitializationScript $CommonFunc -ScriptBlock $ReportTableBlock -ArgumentList $eventlogs,30803,$CaptureDate,([ConsoleColor]'Yellow'),$w
+    $j += Start-Job -name 'SMB Connectivity Error Check - RDMA Connect Failures (Event 30803)' -InitializationScript $CommonFunc -ScriptBlock $ReportTableBlock -ArgumentList $eventlogs,30803,$CaptureDate,([ConsoleColor]'Yellow'),$w,@{'ConnectionType'='=2'}
+
+    $w = @"
+
```
Comment (dl2n) [inline PrivateCloud.DiagnosticInfo/PrivateCloud.DiagnosticInfo.psm1]: you've got crlf problems putting extra newlines from here down. please fix.

File: PrivateCloud.DiagnosticInfo/PrivateCloud.DiagnosticInfo.psm1 [hunk 2]
```diff
@@ -5086,16 +5084,72 @@ WARNING: the SMB Client is receiving RDMA disconnects. This is an error whose ro
...
+Please inspect especially if in the Last5 bucket.
+
+"@
+
+    $j += Start-Job -name 'SMB Connectivity Error Check - TCP/IP Connect Failures (Event 30803)' -InitializationScript $CommonFunc -ScriptBlock $ReportTableBlock -ArgumentList $eventlogs,30803,$CaptureDate,([ConsoleColor]'Yellow'),$w, @{'ConnectionType'='=1'}
+
+    $w = @"
+
+WARNING: the SMB Client is  experiencing long time to complete IO.
```
Comment (dl2n) [inline PrivateCloud.DiagnosticInfo/PrivateCloud.DiagnosticInfo.psm1]: Extra space between 'is experiencing'.

This message in general isn't giving me enough to understand what we're looking at. What's a timeout? Where should we be looking next? Are these 30s operation timeouts? Etc.

File: PrivateCloud.DiagnosticInfo/PrivateCloud.DiagnosticInfo.psm1 [hunk 3]
```diff
@@ -5086,16 +5084,72 @@ WARNING: the SMB Client is receiving RDMA disconnects. This is an error whose ro
...
+
+SMB will close diconnected channels
```
Comment (dl2n) [inline PrivateCloud.DiagnosticInfo/PrivateCloud.DiagnosticInfo.psm1]: "disconnected"

Will each of these show up as disconnected 30804? What's the relationship?

File: PrivateCloud.DiagnosticInfo/PrivateCloud.DiagnosticInfo.psm1 [hunk 4]
```diff
@@ -5086,16 +5084,72 @@ WARNING: the SMB Client is receiving RDMA disconnects. This is an error whose ro
...
+
+"@
+
+    $j += Start-Job -name 'SMB Connectivity Error Check - SMB IO Timeouts (Event 30809)' -InitializationScript $CommonFunc -ScriptBlock $ReportTableBlock -ArgumentList $eventlogs,30809,$CaptureDate,([ConsoleColor]'Yellow'),$w
+
+    $w = @"
+
+WARNING: the SMB Client has disconnect channels because IO took long time.
```
Comment (dl2n) [inline PrivateCloud.DiagnosticInfo/PrivateCloud.DiagnosticInfo.psm1]: "disconnected"

File: PrivateCloud.DiagnosticInfo/PrivateCloud.DiagnosticInfo.psm1 [hunk 5]
```diff
@@ -5086,16 +5084,72 @@ WARNING: the SMB Client is receiving RDMA disconnects. This is an error whose ro
...
+"@
+
+    $j += Start-Job -name 'SMB Connectivity Error Check - SMB Connection Closed (Event 30823)' -InitializationScript $CommonFunc -ScriptBlock $ReportTableBlock -ArgumentList $eventlogs,30823,$CaptureDate,([ConsoleColor]'Yellow'),$w
 
+
+    $eventlogs = (dir $Path\Node_*\System.EVTX).FullName
+
+    $w = @"
+
+WARNING: Dirty Shutdowns has beend detected. Please verify, if those were intended.
```
Comment (dl2n) [inline PrivateCloud.DiagnosticInfo/PrivateCloud.DiagnosticInfo.psm1]: "have been"
no comma between "verify if"

File: PrivateCloud.DiagnosticInfo/PrivateCloud.DiagnosticInfo.psm1 [hunk 6]
```diff
@@ -5086,16 +5084,72 @@ WARNING: the SMB Client is receiving RDMA disconnects. This is an error whose ro
...
+`t There could be connectivity issues due to it.
+
+"@
+
+    $j += Start-Job -name 'Dirty Shutdown Error Check - (Event 41)' -InitializationScript $CommonFunc -ScriptBlock $ReportTableBlock -ArgumentList $eventlogs,41,$CaptureDate,([ConsoleColor]'Yellow'),$w
```
Comment (dl2n) [inline PrivateCloud.DiagnosticInfo/PrivateCloud.DiagnosticInfo.psm1]: Suggest adding a reference to the system log here (e.g., 'System Log Event 41'). The SMB Connectivity events are named for the channel; granted, they could be more specifically the SMB Client Connectivity, but the text helps.

File: PrivateCloud.DiagnosticInfo/PrivateCloud.DiagnosticInfo.psm1 [hunk 7]
```diff
@@ -5744,6 +5798,10 @@ function Show-SddcDiagnosticReport
     # Extract ZIP if neccesary
     $Path = Check-ExtractZip $Path
 
+    # Repror Output file
```
Comment (dl2n) [inline PrivateCloud.DiagnosticInfo/PrivateCloud.DiagnosticInfo.psm1]: No. We don't drop an unsolicited transcript into the current directory. If you'd like to add an option to take a path to tee the output to that would be reasonable. Perhaps if the path is null then autogenerate a path like this ... but it has to be solicited.

Comment (girdharb) [reply PrivateCloud.DiagnosticInfo/PrivateCloud.DiagnosticInfo.psm1]: Looks like, we already transcript in 0_CloudHealthSummary.log, while collecting logs.
Should I just remove the transcript here, and update the "Show-SddcDiagnosticReport -Report Summary -ReportLevel Full $Path" at line 2738 to "Show-SddcDiagnosticReport -Report All -ReportLevel Full $Path"?

File: PrivateCloud.DiagnosticInfo/PrivateCloud.DiagnosticInfo.psm1 [hunk 8]
```diff
@@ -5845,7 +5835,9 @@ function Show-SddcDiagnosticReport
         Write-Output ("Report $r took {0:N2} seconds" -f $td.TotalSeconds)
     }
 
-    Stop-Transcript
+    if(($OutputFile) -and (Test-Path -IsValid $OutputFile)){
```
Comment (dl2n) [inline PrivateCloud.DiagnosticInfo/PrivateCloud.DiagnosticInfo.psm1]: Don't need to Test-Path here

File: PrivateCloud.DiagnosticInfo/PrivateCloud.DiagnosticInfo.psm1 [hunk 9]
```diff
@@ -5798,9 +5788,9 @@ function Show-SddcDiagnosticReport
     # Extract ZIP if neccesary
     $Path = Check-ExtractZip $Path
 
-    # Repror Output file
-    $ReportFileName = (Convert-Path .) + "\" + $Path.Trim('\').Split('\')[-1] + "-SDDCDiagnosticReport.txt"
-    Start-Transcript -Path $ReportFileName -NoClobber
+    if(($OutputFile) -and (Test-Path -IsValid $OutputFile)){
```
Comment (dl2n) [inline PrivateCloud.DiagnosticInfo/PrivateCloud.DiagnosticInfo.psm1]: I don't think -IsValid does what you think it does here. Its just checking for whether a specified drive letter is valid, and beyond that only syntactic plausibility that the remainder could be a path. Its not validating presence of any component of the path in the filesystem. Start-Transcript will do all that.

Should we take a -NoClobber defaulting to $true that a caller could then specify as -NoClobber:$false to allow overwrite? You can call Start-Transcript -NoClobber:$somevalue

**Pull request merged**